<?xml version="1.0" ?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from ./src/experiment_settings/urdf/ur5e_sim.xacro | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="ur5e_robot" xmlns:xacro="http://wiki.ros.org/xacro">
  <gazebo>
    <plugin filename="libgazebo_ros_control.so" name="gazebo_ros_control">
      <!--robotNamespace>/</robotNamespace-->
      <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
      <legacyModeNS>true</legacyModeNS>
      <robotNamespace>/</robotNamespace>
    </plugin>
    <!--
    <plugin name="gazebo_ros_power_monitor_controller" filename="libgazebo_ros_power_monitor.so">
      <alwaysOn>true</alwaysOn>
      <updateRate>1.0</updateRate>
      <timeout>5</timeout>
      <powerStateTopic>power_state</powerStateTopic>
      <powerStateRate>10.0</powerStateRate>
      <fullChargeCapacity>87.78</fullChargeCapacity>     
      <dischargeRate>-474</dischargeRate>
      <chargeRate>525</chargeRate>
      <dischargeVoltage>15.52</dischargeVoltage>
      <chargeVoltage>16.41</chargeVoltage>
    </plugin>
-->
  </gazebo>
  <!--property name="shoulder_height" value="0.089159" /-->
  <!--property name="shoulder_offset" value="0.13585" /-->
  <!-- shoulder_offset - elbow_offset + wrist_1_length = 0.10915 -->
  <!--property name="upper_arm_length" value="0.42500" /-->
  <!--property name="elbow_offset" value="0.1197" /-->
  <!-- CAD measured -->
  <!--property name="forearm_length" value="0.39225" /-->
  <!--property name="wrist_1_length" value="0.093" /-->
  <!-- CAD measured -->
  <!--property name="wrist_2_length" value="0.09465" /-->
  <!-- In CAD this distance is 0.930, but in the spec it is 0.09465 -->
  <!--property name="wrist_3_length" value="0.0823" /-->
  <!-- manually measured -->

  <!-- <link name="root"/>
  <joint name="root2world" type="fixed">
    <parent link="root"/>
    <child link="world"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>
   -->

  <xacro:include filename="$(find ur_description)/urdf/inc/ur_macro.xacro"/>
  <xacro:ur_robot
    prefix=""
    joint_limits_parameters_file="$(find ur_description)/config/ur5e/joint_limits.yaml"
    kinematics_parameters_file="$(find ur_description)/config/ur5e/default_kinematics.yaml"
    physical_parameters_file="$(find ur_description)/config/ur5e/physical_parameters.yaml"
    visual_parameters_file="$(find ur_description)/config/ur5e/visual_parameters.yaml"
    transmission_hw_interface="hardware_interface/PositionJointInterface"
    safety_limits="false"
    safety_pos_margin="0.15"
    safety_k_position="20"
  />

  <link name="ee_link"/>
  <joint name="tool0_ee_link" type="fixed">
    <parent link="tool0"/>
    <child link="ee_link"/>
    <origin xyz="0 0 0" rpy="0 0 0"/>
  </joint>


  <gazebo reference="shoulder_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="upper_arm_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="forearm_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="wrist_1_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="wrist_3_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="wrist_2_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="ee_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  
 
  <material name="camera_aluminum">
    <color rgba="0.5 0.5 0.5 1"/>
  </material>
  <!-- camera body, with origin at bottom screw mount -->
  
  <link name="camera_link">
    <visual>
      <origin rpy="1.5707963267948966 0 1.5707963267948966" xyz="0.0149 -0.0175 0"/>
      <geometry>
        <!-- <box size="${d435_cam_width} ${d435_cam_height} ${d435_cam_depth}"/> -->
        <mesh filename="package://experiment_settings/meshes/d435_sim.dae"/>
        <!--<mesh filename="package://realsense2_description/meshes/d435/d435.dae" />-->
      </geometry>
      <material name="camera_aluminum"/>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 -0.0175 0"/>
      <geometry>
        <box size="0.02505 0.09 0.025"/>
      </geometry>
    </collision>
    <inertial>
      <!-- The following are not reliable values, and should not be used for modeling -->
      <mass value="0.564"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.003881243" ixy="0.0" ixz="0.0" iyy="0.000498940" iyz="0.0" izz="0.003879257"/>
    </inertial>
  </link>
  <!-- camera depth joints and links -->
  <joint name="camera_depth_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="camera_link"/>
    <child link="camera_depth_frame"/>
  </joint>
  <link name="camera_depth_frame"/>
  <joint name="camera_depth_optical_joint" type="fixed">
    <origin rpy="-1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0"/>
    <parent link="camera_depth_frame"/>
    <child link="camera_depth_optical_frame"/>
  </joint>
  <link name="camera_depth_optical_frame"/>
  <!-- camera left IR joints and links -->
  <joint name="camera_left_ir_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0.0 0"/>
    <parent link="camera_depth_frame"/>
    <child link="camera_left_ir_frame"/>
  </joint>
  <link name="camera_left_ir_frame"/>
  <joint name="camera_left_ir_optical_joint" type="fixed">
    <origin rpy="-1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0"/>
    <parent link="camera_left_ir_frame"/>
    <child link="camera_left_ir_optical_frame"/>
  </joint>
  <link name="camera_left_ir_optical_frame"/>
  <!-- camera right IR joints and links -->
  <joint name="camera_right_ir_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 -0.05 0"/>
    <parent link="camera_depth_frame"/>
    <child link="camera_right_ir_frame"/>
  </joint>
  <link name="camera_right_ir_frame"/>
  <joint name="camera_right_ir_optical_joint" type="fixed">
    <origin rpy="-1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0"/>
    <parent link="camera_right_ir_frame"/>
    <child link="camera_right_ir_optical_frame"/>
  </joint>
  <link name="camera_right_ir_optical_frame"/>
  <!-- camera color joints and links -->
  <joint name="camera_color_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0.015 0"/>
    <parent link="camera_depth_frame"/>
    <child link="camera_color_frame"/>
  </joint>
  <link name="camera_color_frame"/>
  <joint name="camera_color_optical_joint" type="fixed">
    <origin rpy="-1.5707963267948966 0 -1.5707963267948966" xyz="0 0 0"/>
    <parent link="camera_color_frame"/>
    <child link="camera_color_optical_frame"/>
  </joint>
  <link name="camera_color_optical_frame"/>
  <gazebo reference="camera_link">
    <self_collide>0</self_collide>
    <enable_wind>0</enable_wind>
    <kinematic>0</kinematic>
    <gravity>1</gravity>
    <!--<mu>1</mu>-->
    <mu2>1</mu2>
    <fdir1>0 0 0</fdir1>
    <!--<slip1>0</slip1>
      <slip2>0</slip2>-->
    <kp>1e+13</kp>
    <kd>1</kd>
    <!--<max_vel>0.01</max_vel>
      <min_depth>0</min_depth>-->
    <sensor name="cameracolor" type="camera">
      <camera name="camera">
        <horizontal_fov>1.2112585008840648</horizontal_fov>
        <image>
          <width>640</width>
          <height>480</height>
          <format>RGB_INT8</format>
        </image>
        <clip>
          <near>0.1</near>
          <far>100</far>
        </clip>
        <noise>
          <type>gaussian</type>
          <mean>0.0</mean>
          <stddev>0.007</stddev>
        </noise>
      </camera>
      <always_on>1</always_on>
      <update_rate>30</update_rate>
      <visualize>1</visualize>
    </sensor>
    <sensor name="cameraired1" type="camera">
      <camera name="camera">
        <horizontal_fov>1.4870205226991688</horizontal_fov>
        <image>
          <width>640</width>
          <height>480</height>
          <format>L_INT8</format>
        </image>
        <clip>
          <near>0.1</near>
          <far>100</far>
        </clip>
        <noise>
          <type>gaussian</type>
          <mean>0.0</mean>
          <stddev>0.05</stddev>
        </noise>
      </camera>
      <always_on>1</always_on>
      <update_rate>90</update_rate>
      <visualize>0</visualize>
    </sensor>
    <sensor name="cameraired2" type="camera">
      <camera name="camera">
        <horizontal_fov>1.4870205226991688</horizontal_fov>
        <image>
          <width>640</width>
          <height>480</height>
          <format>L_INT8</format>
        </image>
        <clip>
          <near>0.1</near>
          <far>100</far>
        </clip>
        <noise>
          <type>gaussian</type>
          <mean>0.0</mean>
          <stddev>0.05</stddev>
        </noise>
      </camera>
      <always_on>1</always_on>
      <update_rate>90</update_rate>
      <visualize>0</visualize>
    </sensor>
    <sensor name="cameradepth" type="depth">
      <camera name="camera">
        <horizontal_fov>1.4870205226991688</horizontal_fov>
        <image>
          <width>640</width>
          <height>480</height>
        </image>
        <clip>
          <near>0.1</near>
          <far>100</far>
        </clip>
        <noise>
          <type>gaussian</type>
          <mean>0.0</mean>
          <stddev>0.100</stddev>
        </noise>
      </camera>
      <always_on>1</always_on>
      <update_rate>90</update_rate>
      <visualize>0</visualize>
    </sensor>
  </gazebo>
  <gazebo>
    <plugin filename="librealsense_gazebo_plugin.so" name="camera">
      <prefix>camera</prefix>
      <depthUpdateRate>30.0</depthUpdateRate>
      <colorUpdateRate>30.0</colorUpdateRate>
      <infraredUpdateRate>30.0</infraredUpdateRate>
      <depthTopicName>depth/image_raw</depthTopicName>
      <depthCameraInfoTopicName>depth/camera_info</depthCameraInfoTopicName>
      <colorTopicName>color/image_raw</colorTopicName>
      <colorCameraInfoTopicName>color/camera_info</colorCameraInfoTopicName>
      <infrared1TopicName>infra1/image_raw</infrared1TopicName>
      <infrared1CameraInfoTopicName>infra1/camera_info</infrared1CameraInfoTopicName>
      <infrared2TopicName>infra2/image_raw</infrared2TopicName>
      <infrared2CameraInfoTopicName>infra2/camera_info</infrared2CameraInfoTopicName>
      <colorOpticalframeName>camera_color_optical_frame</colorOpticalframeName>
      <depthOpticalframeName>camera_depth_optical_frame</depthOpticalframeName>
      <infrared1OpticalframeName>camera_left_ir_optical_frame</infrared1OpticalframeName>
      <infrared2OpticalframeName>camera_right_ir_optical_frame</infrared2OpticalframeName>
      <rangeMinDepth>0.2</rangeMinDepth>
      <rangeMaxDepth>10.0</rangeMaxDepth>
      <pointCloud>True</pointCloud>
      <pointCloudTopicName>depth/color/points</pointCloudTopicName>
      <pointCloudCutoff>0.25</pointCloudCutoff>
      <pointCloudCutoffMax>9.0</pointCloudCutoffMax>
    </plugin>
  </gazebo>
  

  <!-- end of joint list -->
  <!-- colours -->
  <material name="grey">
    <color rgba="0.6627 0.6627 0.6627 1.0"/>
  </material>
  <material name="black">
    <color rgba="0.0 0.0 0.0 1.0"/>
  </material>
  <!-- other elements -->
  <link name="base_cylinder">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <visual>
      <geometry>
        <cylinder length="0.70" radius="0.2"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <colision>
      <geometry>
        <cylinder length="0.70" radius="0.2"/>
      </geometry>
    </colision>
  </link>
  <link name="ground">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <visual>
      <geometry>
        <box size="0.01 3.0 3.0"/>
      </geometry>
      <material name="black"/>
    </visual>
    <colision>
      <geometry>
        <box size="0.01 3.0 3.0"/>
      </geometry>
    </colision>
  </link>
  <link name="ur_gripper">
    <visual>
      <geometry>
        <mesh filename="package://experiment_settings/stl/ur_gripper.stl" scale="0.001 0.001 0.001"/>
      </geometry>
      <material name="grey"/>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://experiment_settings/stl/ur_gripper.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
            <inertial>
            <mass value="0.1" />
            <origin xyz="0 0 0" rpy="0 0 0" />
            <inertia ixx="0.03" iyy="0.03" izz="0.03" ixy="0.0" ixz="0.0" iyz="0.0" />
        </inertial>
  </link>
  <link name="table">
    <visual>
      <geometry>
        <mesh filename="package://experiment_settings/stl/mesa_lab.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </visual>
    <collision>
      <geometry>
        <mesh filename="package://experiment_settings/stl/mesa_lab.stl" scale="0.001 0.001 0.001"/>
      </geometry>
    </collision>
    <inertial>
      <mass value="50"/>
      <inertia ixx="0" ixy="0" ixz="0" iyy="0" iyz="0" izz="0"/>
    </inertial>
  </link>
  <!-- construction -->
  <link name="world"/>
  <joint name="world_ground_joint" type="fixed">
    <parent link="world"/>
    <child link="ground"/>
    <origin rpy="0.0 1.57 0.0" xyz="0.0 0.0 0.0"/>
  </joint>
  <joint name="world_base_joint" type="fixed">
    <parent link="world"/>
    <child link="base_cylinder"/>
    <origin rpy="0.0 0.0 0.0" xyz="0.0 0.0 0.35"/>
  </joint>
  <joint name="base_robot_joint" type="fixed">
    <parent link="base_cylinder"/>
    <child link="base_link"/>
    <origin rpy="0.0 0.0 -1.57" xyz="0.0 0.0 0.35"/>
  </joint>




  <joint name="camera_joint" type="fixed">
    <origin rpy="3.1414 0.0 1.5707" xyz="0.055 0.015 0.03"/>
    <!-- <origin rpy="0.0 -0.0 0.0" xyz="-0.0 -0.0 -0.0"/> -->
    <parent link="ur_gripper"/>
    <child link="camera_bottom_screw_frame"/>
  </joint>
  <link name="camera_bottom_screw_frame"/>
  <joint name="camera_link_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0.0175 0.0125"/>
    <!-- <origin rpy="0.0 -0.0 0.0" xyz="-0.0 -0.0 -0.0"/> -->
    <parent link="camera_bottom_screw_frame"/>
    <child link="camera_link"/>
  </joint>
  <joint name="ur_gripper_joint" type="fixed">
    <parent link="wrist_3_link"/>
    <child link="ur_gripper"/>
    <!-- <origin rpy="1.57 -0.7854 -1.57" xyz="-0.002 -0.08423 -0.15968"/> -->
    <origin rpy="1.57 -0.0 -0.785" xyz="0.08423 0.161 -0.0"/><!-- 0.055 0.175 -->
  </joint>

  <joint name="world_table_joint" type="fixed">
    <parent link="world"/>
    <child link="table"/>
    <origin rpy="-1.57 0.0 -1.57" xyz="-0.2 1.0 0.74"/>
  </joint>   



  <!-- Allegro -->
  <xacro:include filename="$(find allegro_hand_description)/allegro_hand_description_left.xacro"/>
  
  <joint name="palm_2_ur" type="fixed">
    <parent link="ee_link"/>
    <child link="palm_link"/>
    <origin xyz="0.0 0 0.1175" rpy="0 0 -3.92"/> <!-- -0.785-->
  </joint> 

  <transmission name="trans_joint_8">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_8">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_8_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_9">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_9">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_9_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_10">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_10">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_10_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_11">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_11">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_11_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_4">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_4">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_4_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_5">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_5">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_5_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_6">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_6">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_6_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_7">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_7">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_7_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_0">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_0">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_0_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_1">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_1">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_1_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_2">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_2">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_2_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_3">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_3">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_3_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_12">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_12">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_12_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_13">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_13">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_13_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_14">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_14">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_14_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>
    <transmission name="trans_joint_15">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="joint_15">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
        </joint>
        <actuator name="joint_15_motor">
            <hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

</robot >

